src/game_state/Octree.cpp

```diff
@@ -0,0 +1,185 @@
+#include "Octree.h"
+#include <iostream>
+#include <limits>
+#include <sstream>
+#include <cstdlib>
+
+Octree::Octree(
+    std::shared_ptr<std::vector<std::shared_ptr<GameObject>>> objects)
+    : root(constructOctree(objects.get())) {
+  this->objects = objects;
+}
+
+Octree::~Octree() {
+  delete root;
+}
+
+Node* Octree::GetRoot() {
+  return root;
+}
+
+std::shared_ptr<std::vector<std::shared_ptr<GameObject>>> Octree::getObjects() {
+  return objects;
+}
+
+Node* Octree::constructOctree(std::vector<std::shared_ptr<GameObject>>* objs) {
+  // leaf
+  if (objs->size() <= OBJS_IN_LEAF) {
+    glm::vec3 min, max;
+    if (objs->empty()) {
+      std::cerr << "Catastrophic system failure. Unrecoverable" << std::endl;
+      exit(EXIT_FAILURE);
+    }
+    return new Node(objs, AxisAlignedBox(getMin(objs), getMax(objs)),
+                    new std::vector<Node*>());
```

Comment:
I'm going to make an issue for investigating replacing shared_ptrs in some places. We should have a group conversation about how we could change the memory management for GameObjects, specifically if we can say whether or not their lifetime will persist for the entire length of the game. Do you know specifically how shared_ptrs with Nodes could have slowed it down? Was make_shared being called every frame? Or were there a lot of shared_ptr objects being copied around through function calls each frame?